openwisp_notifications/handlers.py

```diff
@@ -158,3 +158,32 @@ def send_email_notification(sender, instance, created, **kwargs):
 )
 def clear_notification_cache(sender, instance, **kwargs):
     Notification.invalidate_cache(instance.recipient)
+
+
+@receiver(pre_delete, dispatch_uid='notification_related_object_deleted')
+def notification_related_object_deleted(sender, instance, **kwargs):
+    instance_id = getattr(instance, 'id', None)
+    instance_content_type = ContentType.objects.get_for_model(instance._meta.model)
+    where = (
+        Q(actor_content_type=instance_content_type)
+        | Q(action_object_content_type=instance_content_type)
+        | Q(target_content_type=instance_content_type)
+    )
+    where = where & (
+        Q(actor_object_id=instance_id)
+        | Q(action_object_object_id=instance_id)
+        | Q(target_object_id=instance_id)
+    )
+    notifications = Notification.objects.filter(where)
```

Comment:
Do you see any problem in just deleting the notifications?
Why do the extra work?
I mean, if a metric is deleted, we can't show notification related to it.
Same for devices or anything else the metric can be related to.

This would then become `Notification.objects.delete(where)` which is far more efficient.

Moreover, this logic could be easily executed in a celery background task, which would help us to keep response times under control.